openedx/frontend-app-course-authoring
# feat: Allow filtering by multiple tags [FC-0040]

## Description

As of https://github.com/openedx/frontend-app-course-authoring/pull/918 , the content search only allows filtering the results by one tag at a time, which is [a limitation of Instantsearch](https://github.com/algolia/instantsearch/issues/1658).

So with this PR, I have replaced our usage of Instantsearch + instant-meilisearch with just direct usage of Meilisearch. Not only does this simplify the code and make our MFE bundle size smaller, but it allows us much more control over how the tags filtering works, so that I can implement searching by multiple tags.

(Trying to modify Instantsearch to do that was too difficult, given the complexity of its codebase).

Related ticket: https://github.com/openedx/modular-learning/issues/201

### Before

* Only one tag can be selected at a time.
* You must select a parent tag before you can select a child filter. The parent tag(s) shows a checkmark even though only the child tag's filter is active.
* You cannot tell if a tag has children or not until you click on it.

![Before](https://github.com/openedx/frontend-app-course-authoring/assets/945577/7eb567b8-a0cd-4ee4-8cf8-4b4a0ad44e00)

### After

* You can select multiple tags. They are filtered using an AND filter.
* You can select a child filter without selecting the parent.
* You can tell which tags have children or not.
* Several dependencies (instantsearch and its depdencies) have been removed.

![After](https://github.com/openedx/frontend-app-course-authoring/assets/945577/c3599485-4a97-4127-92c1-22580b7ecbaa)

## TODO

- [x] Pagination of search results!
- [x] Display a message if tags list may be incomplete - either due to 100 facet limit or 1,000 results keyword limit
- [x] Publish the edx-platform PR

## Testing instructions

1. You must first check out https://github.com/openedx/edx-platform/pull/34534 and run `./manage.py cms reindex_studio --experimental` in your CMS container.
2. Go to the Course Authoring MFE, e.g. http://apps.local.overhang.io:2001/course-authoring/home
3. Optionally, browse to a specific course.
4. Click on the search icon in the header: ![Search Icon](https://github.com/openedx/frontend-app-course-authoring/assets/945577/39918746-37f4-4a7a-82ff-cb418eded05d)
5. Try changing the "Type" filter, and entering keywords in both the overall search keywords field and the tag search keywords fields. Notice how the list of tags dynamically changes to reflect the current results of the search.
6. Try a search that matches a lot of content to check that pagination is working.


## Other information

Private ref: [FAL-3704](https://tasks.opencraft.com/browse/FAL-3704)

Depends on: https://github.com/openedx/frontend-app-course-authoring/pull/918
